packages/react/src/components/cookieConsent/CookieConsentContext.tsx

```diff
@@ -0,0 +1,262 @@
+    current: SupportedLanguage;
+    languageSelectorAriaLabel: string;
+    onLanguageChange: (newLanguage: string) => void;
+  };
+  onAllConsentsGiven?: (consents: ConsentObject) => void;
+  onConsentsParsed?: (consents: ConsentObject, hasUserHandledAllConsents: boolean) => void;
+};
+
+export type CookieConsentContextType = {
+  hasUserHandledAllConsents: () => boolean;
+  content: Content;
+  onAction: CookieConsentActionListener;
+  countApprovedOptional: () => number;
+  areGroupConsentsApproved: (consents: CookieData[]) => boolean;
+};
+
+type CookieConsentContextProps = {
+  cookieDomain?: string;
+  children: React.ReactNode | React.ReactNode[] | null;
+  contentSource: ContentSource;
+};
+
+type CookieConsentAction =
+  | 'approveAll'
+  | 'approveRequired'
+  | 'changeCookieGroupConsents'
+  | 'approveOptional'
+  | 'unapproveOptional'
+  | 'approveSelectedAndRequired';
+
+type CookieConsentActionListener = (action: CookieConsentAction, consents?: string[], value?: boolean) => void;
+
+export const CookieConsentContext = createContext<CookieConsentContextType>({
+  hasUserHandledAllConsents: () => false,
+  content: {} as Content,
+  onAction: () => undefined,
+  countApprovedOptional: () => 0,
+  areGroupConsentsApproved: () => false,
+});
+
+export const getConsentsFromCookieGroup = (groups: CookieGroup[]): ConsentList => {
+  return groups.reduce((ids, currentGroup) => {
+    currentGroup.cookies.forEach((consentData) => {
+      ids.push(consentData.id);
+    });
+    return ids;
+  }, []);
+};
+
+export const Provider = ({ cookieDomain, children, contentSource }: CookieConsentContextProps): React.ReactElement => {
+  const language = contentSource.currentLanguage;
+  const content = useMemo(() => {
+    return createContent(contentSource);
+  }, [language]);
+  const requiredConsents = content.requiredCookies
+    ? getConsentsFromCookieGroup(content.requiredCookies.groups)
+    : undefined;
+  const optionalConsents = content.optionalCookies
+    ? getConsentsFromCookieGroup(content.optionalCookies.groups)
+    : undefined;
+  const consentController = useMemo(
+    () => createConsentController({ requiredConsents, optionalConsents, cookieDomain }),
+    [],
+  );
+  const hasUserHandledAllConsents = () =>
+    consentController.getRequiredWithoutConsent().length === 0 && consentController.getUnhandledConsents().length === 0;
+
+  const mergeConsents = () => ({
+    ...consentController.getRequired(),
+    ...consentController.getOptional(),
+  });
+
+  const [, forceUpdate] = useState<number>(0);
+  const reRender = () => {
+    forceUpdate((p) => p + 1);
+  };
+
+  const save = () => {
+    consentController.save();
+  };
+  const notifyOnAllConsentsGiven = () => {
+    if (content.onAllConsentsGiven && hasUserHandledAllConsents()) {
+      content.onAllConsentsGiven(mergeConsents());
+    }
+  };
+  const update: ConsentController['update'] = (key, value) => {
+    consentController.update(key, value);
```

Comment:
Should we inline this to reduce jumping when reading the code? As it's only one-line I think it could be inlined as well.